src/Grpc.Net.ClientFactory/GrpcClientServiceExtensions.cs

```diff
@@ -306,21 +306,23 @@ public static class GrpcClientServiceExtensions
     {
         ArgumentNullThrowHelper.ThrowIfNull(services);
 
-        services
-            .AddHttpClient(name)
-            .ConfigurePrimaryHttpMessageHandler(() =>
-            {
-                // Set PrimaryHandler to null so we can track whether the user
-                // set a value or not. If they didn't set their own handler then
-                // one will be created by PostConfigure.
-                return null!;
-            });
+        services.AddHttpClient(name);
+
+        // Get PrimaryHandler and store it to compare later. Used to track whether the user set a handler or not.
+        // This action comes before registered user actions so the primary handler here will be the one created by the factory.
+        HttpMessageHandler? initialPrimaryHandler = null;
+        services.Configure<HttpClientFactoryOptions>(name, options =>
+        {
+            options.HttpMessageHandlerBuilderActions.Add(b => initialPrimaryHandler = b.PrimaryHandler);
+        });
```

Comment:
I'd like to discourage the use of `HttpMessageHandlerBuilderActions` if possible... we've already obsoleted `ConfigureHttpMessageHandlerBuilder`

```suggestion
        // Get PrimaryHandler and store it to compare later. Used to track whether the user set a handler or not.
        // This action comes before registered user actions so the primary handler here will be the one created by the factory.
        HttpMessageHandler? initialPrimaryHandler = null;
        
        services
            .AddHttpClient(name)
            .ConfigurePrimaryHttpMessageHandler((primaryHandler, _) =>
            {
                initialPrimaryHandler = primaryHandler;
            });
```

> This action comes before registered user actions so the primary handler here will be the one created by the factory.

Hmmmmmm btw -- it won't be true anymore if the user calls `ConfigureHttpClientDefaults`, right?

(on the other hand, it does _already_ work "unexpectedly" if e.g. the user calls `services.AddHttpClient(name)` manually for some reason, _before_ `AddGrpcHttpClient` 😅)

So:

Couldn't this logic use `ConfigureHttpClientDefaults` instead? or is there smth that prevents it?

I assume, it could be the fact that the user can choose to use _both_ gRPC and "ordinary" HttpClientFactory?

Reply:
Can be related: https://github.com/dotnet/runtime/issues/101719
(e.g. track gRPC client names and apply the defaults only to them 🤔)

Another thought:

What is the reason to _actually_ do this as a post-configure? Is it to avoid creating the handler that won't be needed (if the user ends up creating their own handler)?

if so, there was a draft of an optimization to skip unnecessary primary handler creations if possible: https://github.com/dotnet/runtime/pull/90272/files#diff-282c22a5ee4a798e35d989d1b19b7072d4defc52573b31a876ec440016b4086eR138 -- I can extract it to a separate PR and push it forward. Will that help?

Reply:
grpc-dotnet references 6.0 versions of the packages because the oldest supported .NET Core target is .NET 6. `ConfigureHttpClientDefaults` isn't available.

> What is the reason to actually do this as a post-configure? Is it to avoid creating the handler that won't be needed (if the user ends up creating their own handler)?

Yes. But it isn't important as avoiding unexpected behavior. The code runs once per named client, so the overhead is minimal.